src/api/services/iuf/sessions.go

```diff
@@ -508,13 +522,29 @@ func (s iufService) getDagTasks(session iuf.Session, stageInfo iuf.Stage) ([]v1a
 	if stageInfo.Type == "product" {
 		for _, product := range session.Products {
 			for index, operation := range stageInfo.Operations {
+				if !templateMap[operation.Name] {
+					s.logger.Warnf("The template %v cannot be found in Argo. Make sure you have run upload-rebuild-templates.sh from docs-csm", operation.Name)
+					continue
+				}
+
 				task := v1alpha1.DAGTask{
 					Name: product.Name + "-" + operation.Name,
 				}
 				// dep with a stage
 				if index != 0 {
-					task.Dependencies = []string{
-						product.Name + "-" + stageInfo.Operations[index-1].Name,
+					// check the last dependency is available
+					var lastDependencyOpName string
+					for i := index - 1; i >= 0; i-- {
+						if templateMap[stageInfo.Operations[i].Name] {
+							lastDependencyOpName = stageInfo.Operations[i].Name
+							break
+						}
+					}
+
+					if lastDependencyOpName != "" {
+						task.Dependencies = []string{
```

Comment:
This looks fine for now. In the future, I think we should reconsider whether operations within a stage really have any dependencies on one another. For example, most of the operations in the [deliver-product stage](https://github.com/Cray-HPE/docs-csm/blob/d94bc35885c7feeb8a424ac8b12513f2b2d26be7/workflows/iuf/stages.yaml#L39-L58) could run in parallel. Perhaps we should add something to `stages.yaml` to define whether the operations in a given stage have any dependencies on one another.

Reply:
Yes I was thinking the exact same thing. I think there is some opportunity for improvement here where we can define the dependencies in stages.yaml and construct it with that.

Reply:
Probably worth tracking in a separate Jira, probably even for after the initial release. It's not any worse than it was prior to IUF to do these things serially.

Reply:
Filed [CASMINST-5702](https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-5702)